### PR TITLE
interfaces/apparmor: add missing apparmor rules for journal namespaces

### PR DESCRIPTION
**Pull request by Meulengracht**
Currently the path in apparmor allows writes to the journal socket/stdout, which resides in `/run/systemd/journal`, which is fine for the default journal. However when journal namespaces are used, a new folder will be created for each namespace. Their paths will look like `/run/systemd/journal.{namespace}`

To support journal namespaces, this change is needed

### CODE REVIEW

**Review by mvo5 — APPROVED** [review]
This looks fine to me but it also needs a security review

**Review by alexmurray — CHANGES_REQUESTED** [review]
There is nothing in this PR which specifies what problem there currently is and hence why this change is needed - which makes it hard to assess any possible security tradeoff etc. Can the PR please be updated to provide some background info on this change? Thanks.

**Review by alexmurray — APPROVED** [review]
Thanks for the updated description - do the namespaces follow any particular naming convention? Basically I am wondering can we be stricter than just `*`. Otherwise LGTM.

**Comment by Meulengracht** [comment]
> Thanks for the updated description - do the namespaces follow any particular naming convention? Basically I am wondering can we be stricter than just `*`. Otherwise LGTM.

By inspecting systemd's source code they are pretty generous and allow anything that fits this regex `[a-zA-Z0-9\@:\-\_.\\]+`, we could use that instead of *

**Comment by Meulengracht** [comment]
> > By inspecting systemd's source code they are pretty generous and allow anything that fits this regex [a-zA-Z0-9@:-_.\]+, we could use that instead of *
> 
> aren't we though only interested in the namespace we introduce ourselves? so don't we have a bit more control there?

Correct, if we determine it during runtime we can write a very accurate rule, I have updated this PR to be that, but this is now completely different. I can always revert to the previous approach.

**Review by mvo5 — APPROVED** [review]
Thanks, it's nice that this is super precise now.

**Review by alexmurray — CHANGES_REQUESTED** [review]
Excuse my ignorance, but does the quota name come from a user/admin - ala https://snapcraft.io/docs/quota-groups - and as such, would it be possible for them to try and inject arbitrary apparmor rules via a crafted quota name? I suspect this may be the case, so at a minimum this path should be quoted like so:

```go
template := `  "/run/systemd/%[1]s/socket" w,
  # 'r' shouldn't be needed here, but journald doesn't leak anything so allow
  "/run/systemd/%[1]s/stdout" rw
`
```

On a related note - excuse my ignorance on quota-groups but do we have any constraints on how these are specified/validated?

**Comment by pedronis** [comment]
@alexmurray quota group names should be validated via naming.ValidateQuotaGroup, which has fairly strict rules like for snap names

**Review by pedronis — CHANGES_REQUESTED** [review]
There is a bit of layring confusion and extra complexity in being so precise, as now we in the end had to teach details about journal namespace to the apparmor backend.

**Review by alexmurray — APPROVED** [review]
Thanks for the pointer to `naming.ValidateQuotaGroup` @pedronis - from a security PoV this LGTM but as @pedronis perhaps this is too much of a layering violation now. Either way, I am satisfied that this shouldn't allow admins to inject arbitrary apparmor rules.

File: interfaces/apparmor/backend.go [hunk 1]
```diff
@@ -937,6 +937,27 @@ func (b *Backend) addContent(securityTag string, snapInfo *snap.Info, cmdName st
...
+`
+			// The normal journal log path is /run/systemd/journal/. However,
+			// when a snap is a part of a journal quota group, the snap will need
+			// additional access to a separate path for its journal namespace. This
+			// path will look like /run/systemd/journal.snap-quota-group/. To keep the
+			// allowance of paths minimal, we can determine the exact paths to use
+			// by inspecting the .ExtraLayouts and check for the journal namespace mount.
+			journalRule := "journal"
+			for _, l := range opts.ExtraLayouts {
+				if strings.HasPrefix(l.Bind, path.Join(dirs.SnapSystemdRunDir, "journal.")) {
```
[the remark this answers is not in the document]
Comment (Meulengracht) [reply interfaces/apparmor/backend.go:953]: Yea we prefix namespaces with `snap-`, I'll revert to the previous approach but with the regex

**Review by pedronis — APPROVED** [review]
lgtm, but comment about comments

File: interfaces/apparmor/template.go [hunk 2]
```diff
@@ -191,9 +191,9 @@ var templateCommon = `
 
   # systemd native journal API (see sd_journal_print(4)). This should be in
   # AppArmor's base abstraction, but until it is, include here.
-  /run/systemd/journal/socket w,
-  /run/systemd/journal/stdout rw, # 'r' shouldn't be needed, but journald
-                                  # doesn't leak anything so allow
+  /run/systemd/journal{,.snap-*}/socket w,
+  /run/systemd/journal{,.snap-*}/stdout rw, # 'r' shouldn't be needed, but journald
```
[the remark this answers is not in the document]
Comment (Meulengracht) [reply interfaces/apparmor/template.go:197]: Really good point, thanks! I've updated the PR.

**Pull request merged**